### ISSUE TITLE
Transaction validation

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#1465]
Before you scream "Read the README" I want you to consider this specific case: an application needs to validate (some properties of) a transaction before proceeding and doesn't always has access to full node (and occasionally it's not needed!). E.g. LN clients.

How hard would it be to provide limited validation API that makes some assumptions about UTXOs and such? Sketch of an API:

```rust
/// Builder providing data required for validating transactions
pub struct Validator { /* ... */ }

impl Validator {
    pub fn new() -> Self { /* ... */ }
    pub fn current_block_height(self, height: Height) -> Self { /* ... */ }
    pub fn current_block_time(self, time: Time) -> Self { /* ... */ }
    pub fn dust_fee_rate(self, fee_rate: FeeRate) -> Self { /* ... */ }
    pub fn min_relay_fee_rate(self, fee_rate: FeeRate) -> Self { /* ... */ }
    // Did I forget anything? :)

    // not sure about the input_information parameter - could be some getter closure instead
    pub fn validate(&self, transaction: &Transaction, input_information: &[InputInformation<impl<Borrow<TxOut>>]) -> ValidationReport { /* ... */ }
}

pub struct InputInformation<Out: Borrow<TxOut>> {
    // Does this output belong to a coinbase transaction?
    pub coinbase: bool,
    // If the transaction with this output was confirmed stores height of the block as well as the median time used for lock time checking
    pub confirmed: Option<(Height, Time)>,
    // Previous output
    pub tx_out: Out,
}

#[must_use]
pub struct ValidationReport {
    // indices of signatures that were invalid or missing
    pub invalid_signtures: Vec<usize>,
    pub fee: Amount,
    pub weight: Weight,
    // private fields
}

impl ValidationReport {
    pub fn is_standard(&self) -> bool { /* ... */ }
    pub fn is_valid(&self) -> bool { /* ... */ }
    pub fn spendable_at(&self) -> (Height, Time) { /* ... */ }
    // anything else?
}
```

This should enable some common smart contract patterns such as:

* The transaction is not broadcastable due to time lock but otherwise completely valid
* (some) signatures are missing but everything else is valid

**Comment by apoelstra** [comment]
Yeah, I think these checks are all reasonable. One nit is that the `coinbase` and `tx_out` fields of `InputInformation` should be an enum, since an input has a prevout iff it's not a coinbase.

I'm skeptical that we can check signatures though without a script interpreter. We can certainly add a function that validates a signature against a transaction (maybe taking an optional `SighashCache` for efficiency) but once the signature is in the transaction it's lost to us.

**Comment by sanket1729** [comment]
Concept ACK. I also have an unpublished script interpreter written in rust just for debugging.

ACK on rolling it via `libbitcoinconsensus`. It will be easier to convince people here and get things moving.

I don't mind providing incorrect/buggy error reporting on scripts that are rejected by `libbitcoinconsensus`. I think it will be a painful process to get error reporting merged in the bitcoin core script interpreter. Would really like to have custom error reporting implemented in rust that we have control over. 

I think this might even belong in a separate repo or perhaps bundled with `rust-libbitcoinconsensus`. For people who don't want to double-check the script validation, we can also offer `unchecked` or `dangerous` methods.

The standardness checks are really a pain are they are deliberately not exposed by `libbitcoinconsensus`. Things related to policy/min relay fee can change with different versions of the core software that we need to keep updated.

**Comment by Kixunil** [comment]
> I'm skeptical that we can check signatures though without a script interpreter.

> It'd be really nice if this went in via `libbitcoinconsensus`

Yes, I meant to use `libbitcoinconsensus` for script verification.

> I guess we'd need better error info back from bitcoin core, but that's also doable?

One thing that would be super-awesome is Core being able to tell which data it hashes when signature verification fails. Because if the cause of failed signature is wrong data being hashed it's very hard to debug. My personal experience from past two days. :)

>  unpublished script interpreter written in rust just for debugging.

:heart_eyes: this would be really cool!

> incorrect/buggy error reporting 

Is it really buggy and not just missing information? While not great, missing information is not nearly as bad as wrong message.

>  For people who don't want to double-check the script validation,

I'm not sure what you mean by double-checking but I'd suggest running it through `libbitcoinconsensus` and IFF it returns error while a feature flag is enabled it should run it again via the Rust one to report error. This will be cheap for happy path and people can turn it off if it threatens some kind of DoS attack.

**Comment by tcharding** [comment]
- Labels marked as "API hole" are all marked for inclusion in the next release
- Our target month for the next release is this month.
- This issue is non-trivial to resolve

Can we remove the "API hole" label from this issue and leave this until release 0.31.0?

**Comment by Kixunil** [comment]
I think this is an API hole but given how difficult it is I would make an exception here. It'd be handy for me to have it available in March but I don't think it's realistic.

**Comment by apoelstra** [comment]
Agreed this is too much work for a this-month release.

I think we can do a prerelease in March and can probably get it in.

**Comment by tcharding** [comment]
Cool, added milestone 0.31.0 to signal the comments above.

**Comment by tcharding** [comment]
Removed "API hole" to remove this issue from the ones ear marked for inclusion in 0.30.0 release. Please re-add after release if we want to use this label again for next release.

**Comment by Kixunil** [comment]
It is an API hole, we just make an exception here.
